### PR TITLE
End to end tests with KIND

### PR DESCRIPTION
**Pull request by killianmuldoon**
End to end tests for the prioritize filter and deschedule
strategies of TAS are used here as basic smoke tests to ensure the
scheduler is operating in basic good order as intended. These
tests are designed for basic regression on new changes, but the
framework, based on the Kubernetes in Docker set up, can be used
for more extensive and specific end to end testing.

### CODE REVIEW

File: .github/scripts/e2e_setup_cluster.sh [hunk 1]
```diff
@@ -0,0 +1,158 @@
+#!/bin/bash
```
Comment (togashidm) [inline .github/scripts/e2e_setup_cluster.sh:1]: This is very nice !!

File: .github/e2e/e2e_test.go [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+		want   map[string]bool
+	}{
+		"Label node for deschedule": {policy: deschedule1Policy, want: map[string]bool{"kind-worker2": true}},
+		//"Unlabel node on policy deletion":    {policy: deschedule1_policy, pod: podForPolicy(fmt.Sprintf("pod-%v", rand.String(8)), deschedule1_policy.Name), want:  map[string]bool{"kind-worker2":false}},
+	}
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			res := map[string]bool{}
+			log.Printf("Running: %v\n", name)
+			//defer the running of a cleanup function to remove the policy and pod after the test case
```
Comment (togashidm) [inline .github/e2e/e2e_test.go:98]: Perhaps remove here and add as comment in cleanup function?

File: .github/e2e/e2e_test.go [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+		pod    *v1.Pod
+		want   string
+	}{
+		"Filter all but one node": {policy: filter1Policy, pod: podForPolicy(fmt.Sprintf("pod-%v", time.Now().Unix()), filter1Policy.Name), want: "kind-worker2"},
+		"Filter all nodes":        {policy: filter2Policy, pod: podForPolicy(fmt.Sprintf("pod-%v", rand.String(8)), filter2Policy.Name), want: ""},
+	}
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			log.Printf("Running: %v\n", name)
+			//defer the running of a cleanup function to remove the policy and pod after the test case
```
Comment (togashidm) [inline .github/e2e/e2e_test.go:139]: same as line 98

File: .github/e2e/e2e_test.go [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+		policy *api.TASPolicy
+		pod    *v1.Pod
+		want   string
+	}{
+		"Prioritize to highest score node": {policy: prioritize1Policy, pod: podForPolicy(fmt.Sprintf("pod-%v", time.Now().Unix()), prioritize1Policy.Name), want: "kind-worker2"},
+	}
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			log.Printf("Running: %v\n", name)
+			//defer the running of a cleanup function to remove the policy and pod after the test case
```
Comment (togashidm) [inline .github/e2e/e2e_test.go:174]: same as line 98

**Review by togashidm — COMMENTED** [review]
It is amazing how we can get the TAS up running within KIND just like that !
Huge improvement to support any future e2e test. Thank you.

File: .github/e2e/e2e_test.go [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+	deschedule1Policy = getTASPolicy("deschedule1", deschedule.StrategyType, "deschedule1_metric", "GreaterThan", 8)
+)
+
+func TestTASDeschedule(t *testing.T) {
+	tests := map[string]struct {
+		policy *api.TASPolicy
+		want   map[string]bool
+	}{
+		"Label node for deschedule": {policy: deschedule1Policy, want: map[string]bool{"kind-worker2": true}},
+		//"Unlabel node on policy deletion":    {policy: deschedule1_policy, pod: podForPolicy(fmt.Sprintf("pod-%v", rand.String(8)), deschedule1_policy.Name), want:  map[string]bool{"kind-worker2":false}},
```
[the remark this answers is not in the document]
Comment (killianmuldoon) [reply .github/e2e/e2e_test.go:92]: done

File: .github/e2e/e2e_test.go (hunk 2 above)
Comment (killianmuldoon) [reply .github/e2e/e2e_test.go:98]: I think we should keep this - I want to be really clear on what this means for the flow of the test to make it easier to change down the line.

File: .github/e2e/e2e_test.go [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+
+			nodes, err := cl.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{LabelSelector: labels.Set(lbls.MatchLabels).String()})
+			if err != nil {
+				log.Print(err)
+			}
+			for _, n := range nodes.Items {
+				res[n.Name] = true
+			}
+			if !reflect.DeepEqual(tc.want, res) {
+				//Log extra info if the test fails
```
[the remark this answers is not in the document]
Comment (killianmuldoon) [reply .github/e2e/e2e_test.go:115]: Updated to specify what's been printed out here.

File: .github/e2e/e2e_test.go (hunk 3 above)
Comment (killianmuldoon) [reply .github/e2e/e2e_test.go:139]: Same answer :)

File: .github/e2e/e2e_test.go [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+			log.Print(p.Name)
+			if !reflect.DeepEqual(tc.want, p.Spec.NodeName) {
+				t.Errorf("expected: %v, got: %v", tc.want, p.Spec.NodeName)
+			}
+		})
+	}
+
+}
+
+func TestTASPrioritize(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (killianmuldoon) [reply .github/e2e/e2e_test.go:163]: done

File: .github/e2e/e2e_test.go (hunk 4 above)
Comment (killianmuldoon) [reply .github/e2e/e2e_test.go:174]: same answer :)

**Pull request merged**